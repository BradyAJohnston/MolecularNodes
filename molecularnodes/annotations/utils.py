import inspect
import types
import typing
import bpy
from PIL import Image


def get_all_class_annotations(cls) -> dict:
    """Get all the annotations from the class including base classes"""
    all_annotations = {}
    for base_cls in cls.mro():
        if hasattr(base_cls, "__annotations__"):
            current_annotations = inspect.get_annotations(base_cls, eval_str=True)
            all_annotations.update(current_annotations)
    if "interface" in all_annotations:
        del all_annotations["interface"]
    return all_annotations


def get_blender_supported_type(atype):
    supported_types = (
        str,
        bool,
        int,
        float,
        tuple[float, float],
        tuple[float, float, float],
        tuple[float, float, float, float],
    )
    if atype in supported_types:
        return atype
    if typing.get_origin(atype) is typing.Union or type(atype) is types.UnionType:
        for utype in atype.__args__:
            if utype in supported_types:
                return utype
    return None


def get_view_matrix(obj):
    if obj._render_mode:
        return obj._scene.camera.matrix_world.inverted()
    else:
        return obj._rv3d.view_matrix


def is_perspective_projection(obj):
    if obj._render_mode:
        return obj._scene.camera.data.type != "ORTHO"
    else:
        return obj._rv3d.is_perspective


def render_annotations(
    scene: bpy.types.Scene, render_scale: float, image: Image, image_scale: float
) -> None:
    """Render annotations of all entities to an image"""
    session = scene.MNSession
    session.prune()  # remove any invalid session entities
    for entity in session.entities.values():
        if not hasattr(entity, "annotations"):
            continue
        manager = entity.annotations
        manager._enable_render_mode(scene, render_scale, image, image_scale)
        manager._draw_annotations()
        manager._disable_render_mode()
