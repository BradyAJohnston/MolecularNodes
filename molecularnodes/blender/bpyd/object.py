import bpy
import numpy as np
from typing import Optional
from .attribute import (
    evaluate_object,
    AttributeTypes,
    AttributeType,
    Domains,
    DomainType,
)
from . import attribute
from mathutils import Matrix


class ObjectMissingError(Exception):
    def __init__(self, message):
        self.message = message
        super().__init__(self.message)


class ObjectTracker:
    """
    A context manager for tracking new objects in Blender.

    This class provides a way to track new objects that are added to Blender's bpy.data.objects collection.
    It stores the current objects when entering the context and provides a method to find new objects that were added when exiting the context.

    Methods
    -------
    new_objects():
        Returns a list of new objects that were added to bpy.data.objects while in the context.
    """

    def __enter__(self):
        """
        Store the current objects and their names when entering the context.

        Returns
        -------
        self
            The instance of the class.
        """
        self.objects = list(bpy.context.scene.objects)
        return self

    def __exit__(self, type, value, traceback):
        pass

    def new_objects(self):
        """
        Find new objects that were added to bpy.data.objects while in the context.

        Use new_objects()[-1] to get the most recently added object.

        Returns
        -------
        list
            A list of new objects.
        """
        obj_names = list([o.name for o in self.objects])
        current_objects = bpy.context.scene.objects
        new_objects = []
        for obj in current_objects:
            if obj.name not in obj_names:
                new_objects.append(obj)
        return new_objects

    def latest(self):
        """
        Get the most recently added object.

        This method returns the most recently added object to bpy.data.objects while in the context.

        Returns
        -------
        bpy.types.Object
            The most recently added object.
        """
        return self.new_objects()[-1]


def create_object(
    vertices: np.ndarray | None = None,
    edges: np.ndarray | None = None,
    faces: np.ndarray | None = None,
    name: str = "NewObject",
    collection: bpy.types.Collection | None = None,
) -> bpy.types.Object:
    """
    Create a new Blender object and corresponding mesh.

    Vertices are created for each row in the vertices array. If edges and / or faces are created then they are also
    initialized but default to None.

    Parameters
    ----------
        vertices : np.ndarray, optional
            The vertices of the vertices as a numpy array. Defaults to None.
        edges : np.ndarray, optional
            The edges of the object as a numpy array. Defaults to None.
        faces : np.ndarray, optional
            The faces of the object as a numpy array. Defaults to None.
        name : str, optional
            The name of the object. Defaults to 'NewObject'.
        collection : bpy.types.Collection, optional
            The collection to link the object to. Defaults to None.

    Returns
    -------
        bpy.types.Object
            The created object.
    """
    mesh = bpy.data.meshes.new(name)
    if vertices is None:
        vertices = []
    if edges is None:
        edges = []
    if faces is None:
        faces = []
    mesh.from_pydata(vertices=vertices, edges=edges, faces=faces)
    obj = bpy.data.objects.new(name, mesh)
    if not collection:
        collection = bpy.data.collections["Collection"]
    collection.objects.link(obj)
    return obj


def active_object(context: bpy.types.Context = None) -> bpy.types.Object:
    if context is None:
        return bpy.context.active_object

    return context.active_object


class BlenderObject:
    """
    A convenience class for working with Blender objects
    """

    def __init__(self, obj: bpy.types.Object | None):
        if not isinstance(obj, bpy.types.Object):
            raise ValueError(f"{obj} must be a Blender object of type bpy.types.Object")
        self._object = obj

    @property
    def object(self) -> bpy.types.Object:
        obj = self._object
        if obj is None:
            raise ObjectMissingError(
                "Object is deleted and unable to establish a connection with a new Blender Object."
            )
        return obj

    @object.setter
    def object(self, value: bpy.types.Object) -> None:
        self._object = value

    def store_named_attribute(
        self,
        data: np.ndarray,
        name: str,
        atype: str | AttributeType | None = None,
        domain: str | DomainType = Domains.POINT,
    ) -> None:
        """
        Parameters
        ----------
        data : np.ndarray
            The data to be stored as an attribute.
        name : str
            The name for the attribute. Will overwrite an already existing attribute.
        atype : str or AttributeType or None, optional
            The attribute type to store the data as. Either string or selection from the
            AttributeTypes enum. None will attempt to infer the attribute type from the
            input array.
        domain : str or DomainType, optional
            The domain to store the attribute on. Defaults to Domains.POINT.

        Returns
        -------
        self
        """
        attribute.store_named_attribute(
            self.object, data=data, name=name, atype=atype, domain=domain
        )
        return self

    def evaluate(self):
        return BlenderObject(evaluate_object(self.object))

    def named_attribute(self, name: str, evaluate: bool = False) -> np.ndarray:
        return attribute.named_attribute(self.object, name=name, evaluate=evaluate)

    def transform_origin(self, matrix: Matrix) -> None:
        self.object.matrix_local = matrix * self.object.matrix_world

    def transform_points(self, matrix: Matrix) -> None:
        self.position = self.position * matrix

    @property
    def selected(self) -> np.ndarray:
        return self.named_attribute(".select_vert")

    @property
    def name(self) -> str:
        obj = self.object
        if obj is None:
            return None

        return obj.name

    @name.setter
    def name(self, value: str) -> None:
        obj = self.object
        if obj is None:
            raise ObjectMissingError
        obj.name = value

    @property
    def position(self) -> np.ndarray:
        return self.named_attribute("position")

    @position.setter
    def position(self, value: np.ndarray) -> None:
        self.store_named_attribute(
            value,
            name="position",
            atype=AttributeTypes.FLOAT_VECTOR,
            domain=Domains.POINT,
        )

    def selected_positions(self, mask: Optional[np.ndarray] = None) -> np.ndarray:
        if mask is not None:
            return self.position[np.logical_and(self.selected, mask)]

        return self.position[self.selected]

    def __len__(self) -> int:
        return len(self.object.data.vertices)
