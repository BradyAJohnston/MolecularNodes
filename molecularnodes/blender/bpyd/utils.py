import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)
