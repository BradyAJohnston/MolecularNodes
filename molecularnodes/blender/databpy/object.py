import bpy
import numpy as np
from typing import Union, Optional, Type
from .attribute import (
    store_named_attribute,
    named_attribute,
    Domains,
    DomainType,
)
from mathutils import Matrix


class ObjectTracker:
    """
    A context manager for tracking new objects in Blender.

    This class provides a way to track new objects that are added to Blender's bpy.data.objects collection.
    It stores the current objects when entering the context and provides a method to find new objects that were added when exiting the context.

    Methods
    -------
    new_objects():
        Returns a list of new objects that were added to bpy.data.objects while in the context.
    """

    def __enter__(self):
        """
        Store the current objects and their names when entering the context.

        Returns
        -------
        self
            The instance of the class.
        """
        self.objects = list(bpy.context.scene.objects)
        return self

    def __exit__(self, type, value, traceback):
        pass

    def new_objects(self):
        """
        Find new objects that were added to bpy.data.objects while in the context.

        Use new_objects()[-1] to get the most recently added object.

        Returns
        -------
        list
            A list of new objects.
        """
        obj_names = list([o.name for o in self.objects])
        current_objects = bpy.context.scene.objects
        new_objects = []
        for obj in current_objects:
            if obj.name not in obj_names:
                new_objects.append(obj)
        return new_objects

    def latest(self):
        """
        Get the most recently added object.

        This method returns the most recently added object to bpy.data.objects while in the context.

        Returns
        -------
        bpy.types.Object
            The most recently added object.
        """
        return self.new_objects()[-1]


def create_object(
    vertices: np.ndarray | None = None,
    edges: np.ndarray | None = None,
    faces: np.ndarray | None = None,
    name: str = "NewObject",
    collection: bpy.types.Collection | None = None,
) -> bpy.types.Object:
    """
    Create a new Blender object and corresponding mesh.

    Vertices are created for each row in the vertices array. If edges and / or faces are created then they are also
    initialized but default to None.

    Parameters
    ----------
        vertices : np.ndarray, optional
            The vertices of the vertices as a numpy array. Defaults to None.
        edges : np.ndarray, optional
            The edges of the object as a numpy array. Defaults to None.
        faces : np.ndarray, optional
            The faces of the object as a numpy array. Defaults to None.
        name : str, optional
            The name of the object. Defaults to 'NewObject'.
        collection : bpy.types.Collection, optional
            The collection to link the object to. Defaults to None.

    Returns
    -------
        bpy.types.Object
            The created object.
    """
    mesh = bpy.data.meshes.new(name)
    if vertices is None:
        vertices = []
    if edges is None:
        edges = []
    if faces is None:
        faces = []
    mesh.from_pydata(vertices=vertices, edges=edges, faces=faces)
    obj = bpy.data.objects.new(name, mesh)
    if not collection:
        collection = bpy.data.collections["Collection"]
    collection.objects.link(obj)
    return obj


class BlenderObject:
    """
    A convenience class for working with Blender objects
    """

    def __init__(self, object: bpy.types.Object):
        self.object = object

    def store_named_attribute(
        self,
        data: np.ndarray,
        name: str,
        dtype: Type | None = None,
        domain: str | DomainType = Domains.POINT,
    ) -> None:
        store_named_attribute(self.object, data=data, name=name, domain=domain)

    def named_attribute(self, name: str, evaluate: bool = False) -> np.ndarray:
        return named_attribute(self.object, name=name, evaluate=evaluate)

    def transform_origin(self, matrix: Matrix) -> None:
        self.object.matrix_local = matrix * self.object.matrix_world

    def transform_points(self, matrix: Matrix) -> None:
        self.position = self.position * matrix

    @property
    def selected(self) -> np.ndarray:
        return named_attribute(self.object, ".select_vert")

    @property
    def position(self) -> np.ndarray:
        return named_attribute(self.object, name="position", evaluate=False)

    @position.setter
    def position(self, value: np.ndarray) -> None:
        store_named_attribute(self.object, "position", value)

    def selected_positions(self, mask: Optional[np.ndarray] = None) -> np.ndarray:
        if mask is not None:
            return self.position[np.logical_and(self.selected, mask)]

        return self.position[self.selected]


def bob(object: Union[bpy.types.Object, BlenderObject]) -> BlenderObject:
    """
    Convenience function to convert a Blender object to a BlenderObject
    """
    if isinstance(object, BlenderObject):
        return object
    elif isinstance(object, bpy.types.Object):
        return BlenderObject(object)
    else:
        raise ValueError(
            f"Unknown object type: {object=}"
            "Expected bpy.types.Object or BlenderObject"
        )
