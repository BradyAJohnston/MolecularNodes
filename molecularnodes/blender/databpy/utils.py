import bpy


def evaluate_object(obj: bpy.types.Object):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())
