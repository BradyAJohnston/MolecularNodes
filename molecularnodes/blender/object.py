import bpy
import numpy as np


class ObjectTracker:
    """
    A context manager for tracking new objects in Blender.

    This class provides a way to track new objects that are added to Blender's bpy.data.objects collection.
    It stores the current objects when entering the context and provides a method to find new objects that were added when exiting the context.

    Methods
    -------
    new_objects():
        Returns a list of new objects that were added to bpy.data.objects while in the context.
    """

    def __enter__(self):
        """
        Store the current objects and their names when entering the context.

        Returns
        -------
        self
            The instance of the class.
        """
        self.objects = list(bpy.context.scene.objects)
        return self

    def __exit__(self, type, value, traceback):
        pass

    def new_objects(self):
        """
        Find new objects that were added to bpy.data.objects while in the context.

        Use new_objects()[-1] to get the most recently added object.

        Returns
        -------
        list
            A list of new objects.
        """
        obj_names = list([o.name for o in self.objects])
        current_objects = bpy.context.scene.objects
        new_objects = []
        for obj in current_objects:
            if obj.name not in obj_names:
                new_objects.append(obj)
        return new_objects

    def latest(self):
        """
        Get the most recently added object.

        This method returns the most recently added object to bpy.data.objects while in the context.

        Returns
        -------
        bpy.types.Object
            The most recently added object.
        """
        return self.new_objects()[-1]


def evaluate_object(obj):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())

def create_object(
    vertices: np.ndarray = [],
    edges: np.ndarray = [],
    faces: np.ndarray = [],
    name: str = "NewObject",
    collection: bpy.types.Collection = None,
) -> bpy.types.Object:
    """
    Create a new Blender object, initialised with locations for each vertex.

    If edges and faces are supplied then these are also created on the mesh.

    Parameters
    ----------
        vertices : np.ndarray, optional
            The vertices of the vertices as a numpy array. Defaults to None.
        edges : np.ndarray, optional
            The edges of the object as a numpy array. Defaults to None.
        faces : np.ndarray, optional
            The faces of the object as a numpy array. Defaults to None.
        name : str, optional
            The name of the object. Defaults to 'NewObject'.
        collection : bpy.types.Collection, optional
            The collection to link the object to. Defaults to None.

    Returns
    -------
        bpy.types.Object
            The created object.
    """
    mesh = bpy.data.meshes.new(name)

    mesh.from_pydata(vertices=vertices, edges=edges, faces=faces)

    object = bpy.data.objects.new(name, mesh)

    if not collection:
        # Add the object to the scene if no collection is specified
        collection = bpy.data.collections["Collection"]

    collection.objects.link(object)

    object["type"] = "molecule"

    return object
