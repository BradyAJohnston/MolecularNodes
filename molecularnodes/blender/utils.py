import bpy
from pathlib import Path


def path_resolve(path: str | Path) -> Path:
    if isinstance(path, str):
        return Path(bpy.path.abspath(path))
    elif isinstance(path, Path):
        return Path(bpy.path.abspath(str(path)))
    else:
        raise ValueError(f"Unable to resolve path: {path}")
