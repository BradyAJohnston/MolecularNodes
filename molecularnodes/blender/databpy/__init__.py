from .object import ObjectTracker, BlenderObject, bob, create_object
from .attribute import (
    named_attribute,
    store_named_attribute,
    Attribute,
    AttributeType,
    AttributeTypeInfo,
    AttributeTypes,
    Domains,
    DomainType,
)
