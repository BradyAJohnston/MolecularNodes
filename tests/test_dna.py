import bpy
import pytest
import MDAnalysis as mda
import numpy as np
import molecularnodes as mn
from molecularnodes.entities.trajectory import dna
from databpy.object import LinkedObjectError
from .utils import NumpySnapshotExtension
from .constants import data_dir


class TestOXDNAReading:
    @pytest.fixture(scope="module")
    def file_holl_top(self):
        return data_dir / "oxdna/holliday.top"

    @pytest.fixture(scope="module")
    def file_top_new(self):
        return data_dir / "oxdna/top_new.top"

    @pytest.fixture(scope="module")
    def file_top_new_custom(self):
        return data_dir / "oxdna/top_new_custom.top"

    @pytest.fixture(scope="module")
    def file_top_old(self):
        return data_dir / "oxdna/top_old.top"

    @pytest.fixture(scope="module")
    def file_traj_old_new(self):
        return data_dir / "oxdna/traj_old_new.dat"

    @pytest.fixture(scope="module")
    def file_holl_dat(self):
        return data_dir / "oxdna/holliday.dat"

    @pytest.fixture(scope="module")
    def universe(self, file_holl_top, file_holl_dat):
        return mda.Universe(
            file_holl_top,
            file_holl_dat,
            format=dna.OXDNAReader,
            topology_format=dna.OXDNAParser,
        )

    def test_read_as_universe(self, file_holl_top, file_holl_dat):
        u = mda.Universe(
            file_holl_top,
            file_holl_dat,
            format=dna.OXDNAReader,
            topology_format=dna.OXDNAParser,
        )
        assert u.atoms.n_atoms == 98

    def test_univ_as_traj(self, universe):
        traj = dna.OXDNA(universe)
        assert traj.universe
        with pytest.raises(LinkedObjectError):
            traj.object
        assert all([x in ["A", "C", "T", "G"] for x in traj.res_name])

    def test_univ_snapshot(self, universe: mda.Universe, snapshot_custom):
        traj = dna.OXDNA(universe)
        traj.create_object()
        for name in ["position", "res_name", "res_id", "chain_id"]:
            assert snapshot_custom == getattr(traj, name)

    def test_detect_new_top(self, file_top_old, file_top_new, file_top_new_custom):
        assert dna.OXDNAParser._is_new_topology(file_top_new)
        assert dna.OXDNAParser._is_new_topology(file_top_new_custom)
        assert not dna.OXDNAParser._is_new_topology(file_top_old)

    def test_topo_reading(
        self,
        file_top_old,
        file_top_new,
        file_top_new_custom,
    ):
        top_new = dna.OXDNAParser._read_topo_new(file_top_new)
        top_new_custom = dna.OXDNAParser._read_topo_new(file_top_new_custom)
        top_old = dna.OXDNAParser._read_topo_old(file_top_old)

        for top in [top_new, top_old, top_new_custom]:
            assert top.n_atoms == 12
            assert top.n_residues == 12

    @pytest.mark.parametrize("topfile", ["top_new", "top_new_custom", "top_old"])
    def test_comparing_topologies(self, snapshot, topfile, file_traj_old_new):
        u = mda.Universe(
            data_dir / f"oxdna/{topfile}.top",
            file_traj_old_new,
            topology_format=dna.OXDNAParser,
            format=dna.OXDNAReader,
        )
        traj = dna.OXDNA(u)
        traj.create_object()
        assert len(traj) == 12
        assert snapshot == traj.bonds.tolist()
        for att in ["res_id", "chain_id", "res_name"]:
            assert snapshot == traj.named_attribute(att).tolist()

    def test_reading_example(self):
        traj = dna.OXDNA(
            mda.Universe(
                data_dir / "CanDo2oxDNA/top.top",
                data_dir / "CanDo2oxDNA/traj.oxdna",
                topology_format=dna.OXDNAParser,
                format=dna.OXDNAReader,
            )
        )
        traj.create_object()
        assert len(np.unique(traj.named_attribute("res_id"))) == 15166
        assert len(np.unique(traj.named_attribute("chain_id"))) == 178

    def test_session_register(self, file_holl_top, file_holl_dat):
        session = mn.session.get_session()
        u = mda.Universe(
            file_holl_top,
            file_holl_dat,
            topology_format=dna.OXDNAParser,
            format=dna.OXDNAReader,
        )
        traj = dna.OXDNA(u)
        traj.create_object()

        assert isinstance(session.get(traj.uuid), dna.OXDNA)

    def test_reload_lost_connection(self, snapshot, file_holl_top, file_holl_dat):
        session = mn.session.get_session()
        u = mda.Universe(
            file_holl_top,
            file_holl_dat,
            topology_format=dna.OXDNAParser,
            format=dna.OXDNAReader,
        )
        traj = dna.OXDNA(u)
        traj.create_object()
        obj_name = traj.name
        bpy.context.scene.frame_set(1)
        pos1 = traj.position
        bpy.context.scene.frame_set(2)
        pos2 = traj.position
        assert not np.allclose(pos1, pos2)

        traj_old = session.entities.pop(traj.uuid)

        # the position shouldn't change as we have removed the traj from the session
        bpy.context.scene.frame_set(3)
        pos3 = traj.position
        assert np.allclose(pos2, pos3)
        del traj

        bpy.data.objects[obj_name].select_set(True)
        bpy.ops.mn.reload_trajectory()

        # when reloading the object, a brand new traj had to be created, which updates
        # the uuid on the object, so the old traj will not longer be able to find any
        # matching object and instead we'll have to look back up a new traj based on the
        # the object's uuid
        with pytest.raises(LinkedObjectError):
            traj_old.object.name

        traj = bpy.context.scene.MNSession.get(bpy.context.active_object.uuid)
        assert traj is not None

        pos3 = traj.position

        assert not np.allclose(pos2, pos3)
