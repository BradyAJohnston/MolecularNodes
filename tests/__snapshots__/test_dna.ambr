# serializer version: 1
# name: TestOXDNAReading.test_comparing_topologies[top_new]
  list([
    list([
      1,
      2,
    ]),
    list([
      2,
      3,
    ]),
    list([
      3,
      4,
    ]),
    list([
      4,
      5,
    ]),
    list([
      5,
      6,
    ]),
    list([
      6,
      7,
    ]),
    list([
      7,
      8,
    ]),
    list([
      8,
      9,
    ]),
    list([
      9,
      10,
    ]),
    list([
      10,
      11,
    ]),
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new].1
  list([
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new].2
  list([
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
    5.0,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new].3
  list([
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new_custom]
  list([
    list([
      5,
      6,
    ]),
    list([
      6,
      7,
    ]),
    list([
      7,
      8,
    ]),
    list([
      8,
      9,
    ]),
    list([
      9,
      10,
    ]),
    list([
      10,
      11,
    ]),
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new_custom].1
  list([
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new_custom].2
  list([
    8.0,
    8.0,
    8.0,
    8.0,
    8.0,
    8.0,
    9.0,
    9.0,
    9.0,
    9.0,
    9.0,
    9.0,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_new_custom].3
  list([
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_old]
  list([
    list([
      0,
      1,
    ]),
    list([
      1,
      2,
    ]),
    list([
      2,
      3,
    ]),
    list([
      3,
      4,
    ]),
    list([
      4,
      5,
    ]),
    list([
      6,
      7,
    ]),
    list([
      7,
      8,
    ]),
    list([
      8,
      9,
    ]),
    list([
      9,
      10,
    ]),
    list([
      10,
      11,
    ]),
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_old].1
  list([
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_old].2
  list([
    1.0,
    1.0,
    1.0,
    1.0,
    1.0,
    1.0,
    2.0,
    2.0,
    2.0,
    2.0,
    2.0,
    2.0,
  ])
# ---
# name: TestOXDNAReading.test_comparing_topologies[top_old].3
  list([
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
    42,
  ])
# ---
# name: TestOXDNAReading.test_univ_snapshot
  [[-2.7e-03  3.5e-03  6.1e-03]
   [-2.8e-03  3.7e-03  5.7e-03]
   [-2.8e-03  4.0e-03  5.3e-03]
   [-2.7e-03  4.3e-03  5.0e-03]
   [-2.3e-03  4.6e-03  4.6e-03]
   [-1.9e-03  4.6e-03  4.3e-03]
   [-1.5e-03  4.5e-03  4.0e-03]
   [-1.4e-03  4.4e-03  3.4e-03]
   [-1.4e-03  4.2e-03  3.0e-03]
   [-1.6e-03  4.1e-03  2.5e-03]
   [-1.9e-03  4.2e-03  2.0e-03]
   [-1.6e-03  3.0e-03  9.0e-04]
   [-1.2e-03  2.8e-03  7.2e-04]
   [-7.7e-04  2.8e-03  4.1e-04]
   [-4.2e-04  3.0e-03  8.1e-05]
   [-1.3e-04  3.4e-03 -1.3e-04]
   [ 1.4e-04  3.8e-03 -1.6e-04]
   [ 5.4e-04  4.2e-03 -1.9e-05]
   [ 7.2e-04  4.4e-03  3.7e-04]
   [ 1.1e-03  4.6e-03  7.6e-04]
   [ 1.4e-03  4.5e-03  1.1e-03]
   [ 1.9e-03  4.5e-03  1.2e-03]
   [ 2.1e-03  4.4e-03  5.7e-05]
   [ 1.8e-03  4.0e-03  1.1e-04]
   [ 1.6e-03  3.7e-03  3.2e-04]
   [ 1.2e-03  3.4e-03  5.9e-04]
   [ 7.4e-04  3.4e-03  8.6e-04]
   [ 2.4e-04  3.5e-03  1.0e-03]
   [-1.8e-04  3.7e-03  1.0e-03]
   [-6.2e-04  3.9e-03  8.5e-04]
   [-9.7e-04  4.0e-03  3.9e-04]
   [-1.5e-03  3.8e-03  8.9e-05]
   [-1.7e-03  3.4e-03 -1.7e-04]
   [-2.1e-03  3.0e-03 -2.2e-04]
   [-2.3e-03  2.5e-03 -5.8e-06]
   [-2.7e-03  2.3e-03  3.2e-04]
   [-3.2e-03  2.3e-03  6.7e-04]
   [-3.5e-03  2.5e-03  1.0e-03]
   [-4.0e-03  2.8e-03  1.0e-03]
   [-4.6e-03  2.9e-03  9.7e-04]
   [-5.1e-03  2.9e-03  6.9e-04]
   [-5.4e-03  2.8e-03  3.4e-04]
   [-5.8e-03  2.5e-03  5.3e-05]
   [-6.0e-03  2.0e-03 -1.2e-04]
   [-6.0e-03  2.0e-03  1.1e-03]
   [-5.5e-03  1.8e-03  9.9e-04]
   [-5.1e-03  1.7e-03  7.3e-04]
   [-4.7e-03  1.8e-03  4.0e-04]
   [-4.3e-03  2.1e-03  1.1e-04]
   [-4.1e-03  2.5e-03 -1.5e-04]
   [-3.7e-03  2.9e-03 -2.3e-04]
   [-3.3e-03  3.3e-03 -3.4e-05]
   [-3.0e-03  3.5e-03  3.7e-04]
   [-2.6e-03  3.4e-03  7.2e-04]
   [-2.1e-03  3.2e-03  9.3e-04]
   [-2.1e-03  4.6e-03  1.6e-03]
   [-2.2e-03  5.1e-03  1.4e-03]
   [-2.3e-03  5.5e-03  1.4e-03]
   [-2.1e-03  6.1e-03  1.3e-03]
   [-1.7e-03  6.5e-03  1.2e-03]
   [-1.4e-03  6.7e-03  8.6e-04]
   [-9.9e-04  6.7e-03  5.3e-04]
   [-6.8e-04  6.8e-03  8.2e-05]
   [-6.4e-04  6.8e-03 -4.7e-04]
   [-7.8e-04  6.9e-03 -9.7e-04]
   [-9.4e-04  7.1e-03 -1.4e-03]
   [-1.0e-03  7.6e-03 -1.6e-03]
   [-8.0e-04  8.0e-03 -1.8e-03]
   [-1.6e-03  8.9e-03 -1.9e-03]
   [-1.4e-03  8.2e-03 -2.6e-03]
   [-1.8e-03  8.4e-03 -3.2e-03]
   [-1.4e-03  8.4e-03 -3.1e-03]
   [-9.0e-04  8.3e-03 -2.9e-03]
   [-6.0e-04  8.7e-03 -2.7e-03]
   [-6.2e-04  9.1e-03 -2.5e-03]
   [-5.5e-04  9.3e-03 -2.0e-03]
   [-6.2e-04  7.9e-03 -6.4e-04]
   [-1.0e-03  7.9e-03 -4.1e-04]
   [-1.5e-03  7.5e-03 -3.7e-04]
   [-1.8e-03  7.0e-03 -3.9e-04]
   [-1.8e-03  6.5e-03 -1.8e-04]
   [-1.7e-03  6.0e-03 -3.9e-05]
   [-1.5e-03  5.7e-03  2.3e-04]
   [-1.2e-03  5.6e-03  6.1e-04]
   [-1.1e-03  5.4e-03  1.1e-03]
   [-1.1e-03  5.3e-03  1.6e-03]
   [-1.2e-03  5.2e-03  2.1e-03]
   [-1.5e-03  5.2e-03  2.5e-03]
   [-2.0e-03  5.2e-03  2.8e-03]
   [-2.3e-03  4.9e-03  3.0e-03]
   [-2.6e-03  4.5e-03  3.4e-03]
   [-2.6e-03  4.0e-03  3.7e-03]
   [-2.4e-03  3.6e-03  4.1e-03]
   [-2.1e-03  3.5e-03  4.4e-03]
   [-1.8e-03  3.5e-03  4.8e-03]
   [-1.7e-03  3.7e-03  5.3e-03]
   [-1.7e-03  4.1e-03  5.7e-03]
   [-1.9e-03  4.4e-03  6.1e-03]]
# ---
# name: TestOXDNAReading.test_univ_snapshot.1
  ['C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C'
   'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C'
   'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C'
   'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C'
   'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C'
   'C' 'C' 'C' 'C' 'C' 'C' 'C' 'C']
# ---
# name: TestOXDNAReading.test_univ_snapshot.2
  [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
   1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
   1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
# ---
# name: TestOXDNAReading.test_univ_snapshot.3
  [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
   2 2 2 2 2 2 2 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3
   3 3 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4]
# ---
